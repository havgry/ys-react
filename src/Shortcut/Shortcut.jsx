import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import Anchor, { anchorPropTypes } from '../helpers/Anchor'

const Shortcut = ({
  label,
  icon: Icon,
  className,
  ...props
}) => (
  <Anchor
    className={classNames(
      'ys-shortcut',
      className,
    )}
    {...props}
  >
    <figure
      className="ys-shortcut__icon-container"
      aria-hidden="true"
    >
      <Icon className="ys-shortcut__icon ys-icon" />
    </figure>
    <span className="ys-shortcut__title">
      {label}
    </span>
  </Anchor>
)

Shortcut.propTypes = {
  label: PropTypes.string.isRequired,
  icon: PropTypes.elementType.isRequired,
  href: PropTypes.string,
  className: PropTypes.string,
  ...anchorPropTypes,
}

Shortcut.defaultProps = {
  className: '',
  href: null,
}

export default Shortcut
