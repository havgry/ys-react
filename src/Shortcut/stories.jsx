import React from 'react'
import { storiesOf } from '@storybook/react'
import StoryRouter from 'storybook-react-router'

import { text, select } from '@storybook/addon-knobs'

import Shortcut from './Shortcut'
import { setIcons as icons } from '../../utils/iconMapper'

storiesOf('Shortcut', module)
  .addParameters({ jest: ['Shortcut'] })
  .addDecorator(StoryRouter())
  .add('Anchor element', () => (
    <Shortcut
      label={text('label', 'Musik', 'Content')}
      href={text('href', 'https://yousee.dk', 'Content')}
      icon={icons[select('icon', Object.keys(icons), 'music', 'Visuals')]}
      className={text('className', '', 'Misc')}
    />
  ))
  .add('React Router <NavLink>', () => (
    <Shortcut
      label={text('label', 'Webmail', 'Content')}
      to={text('to', '/dna', 'Content')}
      icon={icons[select('icon', Object.keys(icons), 'mail', 'Visuals')]}
      className={text('className', '', 'Misc')}
    />
  ))
