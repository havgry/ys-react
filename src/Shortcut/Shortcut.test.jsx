import React from 'react'
import { shallow, mount as mountComponent } from 'enzyme'

import Shortcut from './Shortcut'
import dummyIcon from '../../utils/dummyIcon'

const defaultLabel = 'Musik'
const customClassName = 'camel'
const getShortcut = props => (
  <Shortcut
    label={defaultLabel}
    icon={dummyIcon}
    {...props}
  />
)

const shallowRender = props => shallow(getShortcut(props))
const mount = props => mountComponent(getShortcut(props))

describe('<Shortcut>', () => {
  describe('Props', () => {
    it('Throws propType warning if no `label` prop is passed', () => {
      const component = () => shallowRender({
        href: 'https://yousee.dk',
        label: null,
      })
      expect(component).toThrow()
    })

    it('Throws propType warning if no `icon` prop is passed', () => {
      const component = () => shallowRender({
        href: 'https://yousee.dk',
        icon: null,
      })
      expect(component).toThrow()
    })

    it('Has correct defaultProps', () => {
      const component = mount()

      const {
        className,
        href,
        to,
      } = component.props()

      expect(className).toBe('')
      expect(href).toBe(null)
      expect(to).toBe(null)
    })
  })

  describe('Rendering', () => {
    it('Outputs elements with the correct class names', () => {
      const component = shallowRender()

      expect(component.find('figure.ys-shortcut__icon-container').exists()).toBe(true)
      expect(component.find('span.ys-shortcut__title').exists()).toBe(true)
    })

    it('Custom `className` is being applied to the container', () => {
      const component = shallowRender({
        className: customClassName,
      })

      expect(component.find(`.ys-shortcut.${customClassName}`).exists()).toBe(true)
    })

    it('Passes other generic props to container element', () => {
      const component = shallowRender({
        id: 'test',
        tabIndex: 0,
        className: 'owl',
      })

      expect(component.props().id).toBe('test')
      expect(component.props().tabIndex).toBe(0)
      expect(component.props().className).toBe('ys-shortcut owl')
    })
  })

  describe('Snapshots', () => {
    it('With custom `className`', () => {
      const component = shallowRender({
        href: 'https://yousee.dk',
        className: customClassName,
      })

      expect(component).toMatchSnapshot()
    })

    it('Without custom `className`', () => {
      const component = shallowRender({
        href: 'https://yousee.dk',
      })

      expect(component).toMatchSnapshot()
    })
  })
})
