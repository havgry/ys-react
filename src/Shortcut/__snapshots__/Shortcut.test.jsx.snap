// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Shortcut> Snapshots With custom \`className\` 1`] = `
<Anchor
  className="ys-shortcut camel"
  disabled={false}
  href="https://yousee.dk"
  to={null}
>
  <figure
    aria-hidden="true"
    className="ys-shortcut__icon-container"
  >
    <dummyIcon
      className="ys-shortcut__icon ys-icon"
    />
  </figure>
  <span
    className="ys-shortcut__title"
  >
    Musik
  </span>
</Anchor>
`;

exports[`<Shortcut> Snapshots Without custom \`className\` 1`] = `
<Anchor
  className="ys-shortcut"
  disabled={false}
  href="https://yousee.dk"
  to={null}
>
  <figure
    aria-hidden="true"
    className="ys-shortcut__icon-container"
  >
    <dummyIcon
      className="ys-shortcut__icon ys-icon"
    />
  </figure>
  <span
    className="ys-shortcut__title"
  >
    Musik
  </span>
</Anchor>
`;
